### PR TITLE
Add LayoutLMv2ForRelationExtraction

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the relation extraction head of LayoutLMv2, which was a highly requested feature as seen in #14330 #15451 #18091

### CODE REVIEW

**Comment by NielsRogge** [comment]
@sgugger I'm getting the following error from make fixup:

```
Checking all objects are properly documented.
Traceback (most recent call last):
  File "/home/niels/python_projects/transformers/utils/check_repo.py", line 788, in <module>
    check_repo_quality()
  File "/home/niels/python_projects/transformers/utils/check_repo.py", line 782, in check_repo_quality
    check_all_objects_are_documented()
  File "/home/niels/python_projects/transformers/utils/check_repo.py", line 693, in check_all_objects_are_documented
    raise Exception(
Exception: The following objects are in the public init so should be documented:
 - LayoutLMv2ForRelationExtraction
```

However, this model is added to layoutlmv2.mdx, so not sure why this error occurs.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 1]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+            pred_relations.append(rel)
+        return pred_relations
+
+    def forward(self, hidden_states, entities, relations):
+        batch_size, max_n_words, context_dim = hidden_states.size()
+        device = hidden_states.device
+        relations, entities = self.build_relation(relations, entities)
+        loss = 0
+        all_pred_relations = []
+        for b in range(batch_size):
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1572]: Same here, consider b -> idx

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 2]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+            head_entities = torch.tensor(relations[b]["head"], device=device)
+            tail_entities = torch.tensor(relations[b]["tail"], device=device)
+            relation_labels = torch.tensor(relations[b]["label"], device=device)
+            entities_start_index = torch.tensor(entities[b]["start"], device=device)
+            entities_labels = torch.tensor(entities[b]["label"], device=device)
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1577]: It's up to the user to provide tensors as inputs, and on the right device, this should just extract the keys.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 3]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+            head_label_repr = self.entity_emb(head_label)
+
+            tail_index = entities_start_index[tail_entities]
+            tail_label = entities_labels[tail_entities]
+            tail_label_repr = self.entity_emb(tail_label)
+
+            head_repr = torch.cat(
+                (hidden_states[b][head_index], head_label_repr),
+                dim=-1,
+            )
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1589]:
```suggestion
            head_repr = torch.cat((hidden_states[b][head_index], head_label_repr), dim=-1)
```
Fits in one line.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 4]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+            tail_repr = torch.cat(
+                (hidden_states[b][tail_index], tail_label_repr),
+                dim=-1,
+            )
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1593]:
```suggestion
            taill_repr = torch.cat((hidden_states[b][tail_index], tail_label_repr), dim=-1)
```

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 5]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+            nn.Linear(config.hidden_size, config.hidden_size // 2),
+            nn.ReLU(),
+            nn.Dropout(config.hidden_dropout_prob),
+        )
+        self.ffnn_head = copy.deepcopy(projection)
+        self.ffnn_tail = copy.deepcopy(projection)
+        self.rel_classifier = BiaffineAttention(config.hidden_size // 2, 2)
+        self.loss_fct = CrossEntropyLoss()
+
+    def build_relation(self, relations, entities):
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1519]: This looks like a preprocessing method. It shouldn't be part of the model.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 6]
```diff
@@ -1424,3 +1462,249 @@ def forward(
...
+        >>> model = LayoutLMv2ForRelationExtraction.from_pretrained("microsoft/layoutlmv2-base-uncased")
+
+        >>> dataset = load_dataset("hf-internal-testing/fixtures_docvqa")
+        >>> image_path = dataset["test"][0]["file"]
+        >>> image = Image.open(image_path).convert("RGB")
+        >>> encoding = processor(image, return_tensors="pt")
+
+        >>> # instantiate relations as empty at inference time
+        >>> encoding["entities"] = [{"start": [0, 4], "end": [3, 6], "label": [2, 1]}]
+        >>> encoding["relations"] = [{"start_index": [], "end_index": [], "head": [], "tail": []}]
```
Comment (sgugger) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:1666]: Model cant take lists as inputs as they then won't work with ONNX/distributed etc. This should all be tensors.

File: tests/models/layoutlmv2/test_modeling_layoutlmv2.py [hunk 7]
```diff
@@ -485,6 +572,230 @@ def test_initialization(self):
...
+                    if p1.data.ne(p2.data).sum() > 0:
+                        models_equal = False
+
+            self.assertTrue(models_equal)
+
+            # Avoid memory leak. Without this, each call increase RAM usage by ~20MB.
+            # (Even with this call, there are still memory leak by ~0.04MB)
+            self.clear_torch_jit_class_registry()
+
+    # overwrite as LayoutLMv2ForRelationExtraction outputs dictonaries containing integers rather than tensors
```
Comment (sgugger) [inline tests/models/layoutlmv2/test_modeling_layoutlmv2.py:722]: Which really shouldn't be the case...

**Review by sgugger — COMMENTED** [review]
Thanks for the PR, but really not convinced by the model added as it is. Some preprocessing code is part of the model, which is not how we usually do things in Transformers. Some inputs are not tensors, which causes multiple problems (tests overridden are just the tip of the iceberg) as a result.

The whole preprocessing should be added as a method on the preprocessing class of LayoutLMv2 and the model should accept tensors, listed as keyword arguments and not inside a dictionary.

**Comment by lamaeldo** [comment]
Hey @NielsRogge , any chance that this will ever be implemented?
Looking around the history of the PR and Iissues, it seems that there was a fair bit of interest

**Comment by NielsRogge** [comment]
Hi @lamaeldo,

The reason the PR wasn't merged is because models need to output fixed size tensors, to make sure things like distributed training and ONNX export work. However LayoutLMv2ForRelationExtraction outputs lists of tensors in its current implementation, due to each example in the batch having a different amount of relations. So we would need to pad them up to a fixed size such that the model outputs fixed size tensors.

Haven't looked into that yet but if you're willing to contribute, let me know!

Btw I do have a notebook on fine-tuning this model [here](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/LayoutXLM).

**Pull request closed without merging**